### PR TITLE
Goal response callback compatibility shim with deprecation of old signature

### PR DESCRIPTION
**Pull request by ivanpauno**
https://github.com/ros2/rclcpp/pull/1311 broke the goal response callback signature, this adds a compatibility shim so we affect less users in the next galactic release (see [this comment](https://github.com/ros2/rclcpp/pull/1311#issuecomment-728841765)).

I think this is a good idea, but there's a minor caveat: `goal_response_callback` was before a `std::function<>`, with the full `std::function` API available, and now there are a few things that the shim type doesn't implement (e.g.: `assign`, `swap`), though we could try to implement all of them.

### CODE REVIEW

**Review by hidmic — APPROVED** [review]
Code overall looks OK, pending green CI.

File: rclcpp_action/include/rclcpp_action/client.hpp [hunk 1]
```diff
@@ -268,13 +268,105 @@ class Client : public ClientBase
...
+    GoalResponseCallback &
+    operator=(NewSignature new_callback) {new_callback_ = std::move(new_callback); return *this;}
+
+    void
+    operator()(typename GoalHandle::SharedPtr goal_handle) const {
+      if (RCUTILS_LIKELY(static_cast<bool>(new_callback_))) {
+        new_callback_(std::move(goal_handle));
+        return;
+      }
+      if (static_cast<bool>(old_callback_)) {
```
[the remark this answers is not in the document]
Comment (ivanpauno) [reply rclcpp_action/include/rclcpp_action/client.hpp]: Fixed in https://github.com/ros2/rclcpp/pull/1495/commits/d838e928a1cd1cb3c941232f2af1cb5959ed23a4.

> @ivanpauno IIRC std::function bool operator is explicit. There should be no need for a static_cast in an if clause.

Yes, I was using `RCUTILS_LIKELY()` in some places, and with that it wasn't working (gcc `__builtin_expect()` takes an `int`, and the "contextual conversion" isn't allowed there).

PS: By the way, if the bool operator were implicit it should also be no need of a `static_cast<bool>()` in an if clause.
  I didn't know, but in the "condition" of an `if`, `for`, `etc`; "[contextual conversions](https://en.cppreference.com/w/cpp/language/implicit_conversion#Contextual_conversions)" are allowed (which is kind of an implicit explicit cast :joy:). `cpp` keeps surprising me every day, I always thought those `operator bool()` definitions were implicit.

File: rclcpp_action/include/rclcpp_action/client.hpp [hunk 2]
```diff
@@ -268,13 +268,105 @@ class Client : public ClientBase
...
+        return;
+      }
+      if (RCUTILS_LIKELY(static_cast<bool>(new_callback_))) {
+        new_callback_(std::move(goal_handle_future).get_future().share());
+        return;
+      }
+      throw std::bad_function_call{};
+    }
+
+    operator bool() const noexcept {
```
[the remark this answers is not in the document]
Comment (ivanpauno) [reply rclcpp_action/include/rclcpp_action/client.hpp]: Done in https://github.com/ros2/rclcpp/pull/1495/commits/8e9b87d7af23154c723f59ea9c5d1212a1c1a55c

**Comment by ivanpauno** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13304/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8237/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11037/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13347/)

**Review by hidmic — APPROVED** [review]
LGTM !

**Comment by jacobperron** [comment]
Thanks, LGTM too!

**Pull request merged**